h2o-algos/src/main/java/hex/kmeans/KMeansSimplexSolver.java

```diff
@@ -374,56 +345,82 @@ public Frame assignClusters() {
+                    sources.set(i * _secondLayerSize + j, i);
+                    targets.set(i * _secondLayerSize + j, numberOfPoints + j);
                 }
             } else {
                 if (i < _nodeSize - 1) {
-                    _sources.set(numberOfPoints* _secondLayerSize +i-numberOfPoints, i);
-                    _targets.set(numberOfPoints* _secondLayerSize +i-numberOfPoints, _nodeSize - 1);
+                    sources.set(numberOfPoints* _secondLayerSize +i-numberOfPoints, i);
+                    targets.set(numberOfPoints* _secondLayerSize +i-numberOfPoints, _nodeSize - 1);
                 }
             }
         }
-
+        
         for (long i = 0; i < _nodeSize; i++) {
             long demand = demands.at8(i);
             if (demand >= 0) {
-                _sources.set(_edgeSize + i, _nodeSize);
-                _targets.set(_edgeSize + i, i);
+                sources.set(_edgeSize + i, _nodeSize);
+                targets.set(_edgeSize + i, i);
             } else {
-                _sources.set(_edgeSize + i, i);
-                _targets.set(_edgeSize + i, _nodeSize);
+                sources.set(_edgeSize + i, i);
+                targets.set(_edgeSize + i, _nodeSize);
             }
             if (i < _nodeSize - 1) {
                 _nextDepthFirst.set(i, i + 1);
             }
-            _edgeFlow.set(_edgeSize +i, Math.abs(demand));
+            
+            _edgeFlowRest.set(_secondLayerSize+i, Math.abs(demand));
+            
             _nodePotentials.set(i, demand < 0 ? maxCapacity : -maxCapacity);
             _parents.set(i, _nodeSize);
             _parentEdges.set(i, i + _edgeSize);
             _previousNodes.set(i, i - 1);
             _lastDescendants.set(i, i);
         }
+        
         _parents.set(_nodeSize, -1);
         _subtreeSize.set(_nodeSize, _nodeSize + 1);
         _nextDepthFirst.set(_nodeSize - 1, _nodeSize);
         _previousNodes.set(0, _nodeSize);
         _previousNodes.set(_nodeSize, _nodeSize - 1);
         _lastDescendants.set(_nodeSize, _nodeSize - 1);
+        
+        _sourcesReader = sources.new Reader();
+        _targetsReader = targets.new Reader();
     }
 
+    /**
+     * Check if tree is feasible and the algorithm can stop.
+     * @return
+     */
     public boolean isInfeasible() {
-        for(long i = _edgeFlow.length() - _secondLayerSize + 1; i < _edgeFlow.length(); i++) {
-            if(_edgeFlow.at8(i) > 0){
+        for(long i = 1; i < _secondLayerSize + 2; i++) {
+            if(_edgeFlowRest.at8(_edgeFlowRest.length() - i) > 0) {
                 return true;
             }
         }
         return false;
     }
 
+    /**
+     * Check if the constraints are satisfied. 
+     * If yes, the algorithm can continue as standard K-means and save time. Useful when constraints are small numbers
+     * @return 
+     */
+    public boolean areConstraintsSatisfied() {
+        for(long i = 2; i < _secondLayerSize + 2; i++) {
```

Comment:
What is the size of the layer? Sequential reads benefit the most from using Vec.Reader.

Reply:
Good point. The size of layer equals `k`. It is not usually a huge number. However, the reader can help. Fixed.